### PR TITLE
Closes #3173: handle variables with certain keywords

### LINKED ISSUE #3173: Variables with uppercase characters get mangled
## Bug Report

**What did you do?**
Created a CR with following spec:
```
spec:
  egressIPs:
    - 10.1.1.1
    - 4.2.2.2
```

**What did you expect to see?**
following ansible variable:
```
"egressIPs":[
            "10.1.1.1",
            "4.2.2.2"
        ],
```

**What did you see instead? Under which circumstances?**
following ansible variable:
```
"egress_i_ps":[
            "10.1.1.1",
            "4.2.2.2"
        ],
```

**Environment**
* operator-sdk version:
operator-sdk version: "v0.18.0", commit: "921c33982687d6e9219fe0a1a326515fb37ae040", kubernetes version: "v1.18.2", go version: "go1.13.10 linux/amd64"

* go version:

* Kubernetes version information:

* Kubernetes cluster kind: 

* Are you writing your operator in ansible, helm, or go? ansible

**Possible Solution**

**Additional context**

### PR DESCRIPTION
**Pull request by Venkat19967**
Added the changes for plural in paramconv.go

**Description of the change:**
This change is to fix the issue :
https://github.com/operator-framework/operator-sdk/issues/3173

I have made changes to the file such that the error case mentioned in the issue is resolved
the changes have been made to ToSnake(). 
Added a few more test cases to check all the edge cases 

The camelCase is working fine as is. 

**Motivation for the change:**

### CODE REVIEW

File: pkg/ansible/paramconv/paramconv.go [hunk 1]
```diff
@@ -84,6 +91,38 @@ func ToSnake(s string) string {
...
+						s = s[0:i+1] + "S" + "_" + s[i+2:]
+					} else {
+						s = s[0:i-count+1] + "_" + s[i-count+1:i+1] + "S" + "_" + s[i+2:]
+					}
+				}
+
+			}
+		}
+
+	}
```
Comment (jmrodri) [inline pkg/ansible/paramconv/paramconv.go:125]: @Venkat19967 My first thought initially was wondering if there was a way to combine this loop with the one below, but I see you are actually modifying the string `s` that is used in the loop below. Given that I would break this out into a function, something like `preprocessWordMappings(...) string`. Not sure what parameters it would take but it should return a string.

So then in this method the above code would be replaced by the method call:

`s = preprocessWordMappings(...)`

I think that will make it easier to understand. Then add a comment above explaining what you are doing in the method.

File: pkg/ansible/paramconv/paramconv_test.go [hunk 2]
```diff
@@ -117,6 +117,11 @@ func TestToCamel(t *testing.T) {
 			args: args{"var_k8s"},
 			want: "varK8s",
 		},
+		{
+			name: "zeus",
```
Comment (jmrodri) [inline pkg/ansible/paramconv/paramconv_test.go:121]: change `zeus` to be `should handle special words`

File: pkg/ansible/paramconv/paramconv_test.go [hunk 3]
```diff
@@ -181,6 +186,11 @@ func TestToSnake(t *testing.T) {
 			args: args{"this_should_be_a_noop"},
 			want: "this_should_be_a_noop",
 		},
+		{
+			name: "zeus",
```
Comment (jmrodri) [inline pkg/ansible/paramconv/paramconv_test.go:190]: change `zeus` to be `should handle special words`

File: pkg/ansible/paramconv/paramconv.go [hunk 4]
```diff
@@ -17,6 +17,8 @@
 package paramconv
 
 import (
+	// "fmt"
+
```
Comment (jmrodri) [inline pkg/ansible/paramconv/paramconv.go:21]: definitely remove all the commented out code.

File: pkg/ansible/paramconv/paramconv.go [hunk 5]
```diff
@@ -72,10 +75,14 @@ func ToCamel(s string) string {
...
 func ToSnake(s string) string {
+	// fmt.Printf("\n#################################################\n")
+	// fmt.Printf("\nInside to snake\n")
 	s = addWordBoundariesToNumbers(s)
 	s = strings.Trim(s, " ")
+	// fmt.Printf("\nThe Value of s = %s\n", s)
 	var prefix string
 	char1 := []rune(s)[0]
+	// fmt.Printf("\nThe Value of rune = %c\n", char1)
 	if char1 >= 'A' && char1 <= 'Z' {
```
Comment (jmrodri) [inline pkg/ansible/paramconv/paramconv.go:86]: I know this is not your code, but please add a comment above this code:

```
// prepending an underscore (_) if the word begins with a Capital Letter
```

File: pkg/ansible/paramconv/paramconv.go [hunk 6]
```diff
@@ -84,6 +91,38 @@ func ToSnake(s string) string {
 	bits := []string{}
 	n := ""
 	iReal := -1
+	count := 0
+	length := len(s)
+	for i, v := range s {
+		if i+1 < length {
+			next := s[i+1]
+			count = count + 1
```
Comment (jmrodri) [inline pkg/ansible/paramconv/paramconv.go:99]: change `count = count +` to `count++`

File: pkg/ansible/paramconv/paramconv.go [hunk 7]
```diff
@@ -84,6 +91,38 @@ func ToSnake(s string) string {
...
+					}
+					// fmt.Printf("This is the final string: %s", s)
+				} else if next >= 'a' && next <= 'z' && next == 's' {
+					if _, ok := wordMapping[strings.ToLower(s[i-count+1:i+2])]; ok && i+2 == length {
+						s = s[0:i-count+1] + "_" + s[i-count+1:i+1] + "S"
+						// fmt.Printf("\n this si the S : %s\n", s)
+					} else if _, ok := wordMapping[strings.ToLower(s[i-count+1:i+2])]; ok && i-count+1 == 0 {
+						s = s[0:i+1] + "S" + "_" + s[i+2:]
+					} else {
+						s = s[0:i-count+1] + "_" + s[i-count+1:i+1] + "S" + "_" + s[i+2:]
```
Comment (jmrodri) [inline pkg/ansible/paramconv/paramconv.go:118]: I don't like the concatenations but I don't know of a better way. 
Using `fmt.Sprintf` seems to be about the same complexity.

`s = fmt.Sprintf("%s_%sS_%s", s[0:i-count+1], s[i-count+1:i+1], s[i+2:])`

File: pkg/ansible/paramconv/paramconv.go [hunk 8]
```diff
@@ -70,6 +71,41 @@ func ToCamel(s string) string {
...
+//This function modifies the string to handle special words that are a part of wordMapping
+//Add values to handle wider number of cases
+func preprocessWordMapping(s string) string {
+	count := 0
+	length := len(s)
+	for i, v := range s {
+		if i+1 < length {
+			next := s[i+1]
+			count++
+			if (v >= 'A' && v <= 'Z' && next >= 'a' && next <= 'z') || (v >= 'a' && v <= 'z' && next >= 'A' && next <= 'Z') {
```
Comment (jmrodri) [inline pkg/ansible/paramconv/paramconv.go:83]: I don't like the deep indentation. What if we flip the logic of the if. So let's do this instead:

```
for i, v := range s {
    if i+1 > length {
        continue
    }

    next := s[i+1]
    count++
    ...
}
```

Doing this the code won't have to be nested so deep for the one edge case where `i+1 > length`.

File: pkg/ansible/paramconv/paramconv.go [hunk 9]
```diff
@@ -70,6 +71,41 @@ func ToCamel(s string) string {
 	return ret
 }
 
+//This function modifies the string to handle special words that are a part of wordMapping
+//Add values to handle wider number of cases
```
Comment (jmrodri) [inline pkg/ansible/paramconv/paramconv.go:75]: Pet peeve of mine, add a space after the `//` :)

File: pkg/ansible/paramconv/paramconv.go [hunk 10]
```diff
@@ -28,6 +28,7 @@ var (
 		"http": "HTTP",
```
Comment (camilamacedo86) [inline pkg/ansible/paramconv/paramconv.go:28]: Missing fragment

File: pkg/ansible/paramconv/paramconv.go [hunk 11]
```diff
@@ -70,6 +71,42 @@ func ToCamel(s string) string {
 	return ret
 }
 
+// This function modifies the string to handle special words that are a part of wordMapping
```
Comment (camilamacedo86) [inline pkg/ansible/paramconv/paramconv.go:74]: This one is not very readable. I am not sure if could not have an easy way/more understandable way to achieve this goal. However, could we improve the comments here to make clear that it is used to not change/snake the special worlds in the wordMapping? Could we add in the comment an example over the scenario and the expected result?

File: pkg/ansible/paramconv/paramconv.go [hunk 12]
```diff
@@ -70,6 +71,42 @@ func ToCamel(s string) string {
...
+// Add values to handle wider number of cases
+func preprocessWordMapping(s string) string {
```
Comment (camilamacedo86) [inline pkg/ansible/paramconv/paramconv.go:76]: Also, could we please add an explanation over all rules applied here?

File: pkg/ansible/paramconv/paramconv.go [hunk 13]
```diff
@@ -70,6 +71,42 @@ func ToCamel(s string) string {
...
+	count := 0
+	length := len(s)
+	for i, v := range s {
+		if i+1 >= length {
+			break
+		}
+		next := s[i+1]
+		count++
+		if (v >= 'A' && v <= 'Z' && next >= 'a' && next <= 'z') || (v >= 'a' && v <= 'z' && next >= 'A' && next <= 'Z') {
```
Comment (tengqm) [inline pkg/ansible/paramconv/paramconv.go:85]: for readability's sake, maybe add some util functions for checking if a character is upper or lower cased?

File: changelog/fragments/3173-bugfix.yaml [hunk 14]
```diff
@@ -0,0 +1,5 @@
+entries:
+  - description: >
+      Fixed mangling of Ansible-based variables that have special words in all upper case
+    kind: bugfix
+    breaking: false
```
Comment (camilamacedo86) [inline changelog/fragments/3173-bugfix.yaml:5]: missing a space at the end of the file

Comment (camilamacedo86) [inline changelog/fragments/3173-bugfix.yaml:5]: it is  count as description when shold not. see the template.

File: pkg/ansible/paramconv/paramconv.go [hunk 15]
```diff
@@ -70,6 +70,49 @@ func ToCamel(s string) string {
 	return ret
 }
 
+// This function modifies the string to handle special words that are a part of wordMapping
+/* In this function we process the input string and find the occurrence of the values in wordMapping
+   and we append "_" on both sides of that word so that it is considered to be one word*/
+// We even handle plurals of the values in wordMapping
+// If we find "s" i.e plural, we capitalize it so that it is considered to be a part of same abbreviation
+
```
Comment (estroz) [inline pkg/ansible/paramconv/paramconv.go:78]:
```suggestion
```

File: changelog/fragments/3173-bugfix.yaml [hunk 16]
```diff
@@ -0,0 +1,5 @@
+entries:
+  - description: >
+      Fixed mangling of Ansible-based variables that have special words in all upper case
```
Comment (camilamacedo86) [inline changelog/fragments/3173-bugfix.yaml:3]:
```suggestion
handle variables with certain keywords
      Fix Ansible based-Operators for not mangled variables with uppercase characters when they have special words (e.g IP).
```

File: pkg/ansible/paramconv/paramconv.go [hunk 17]
```diff
@@ -70,6 +70,49 @@ func ToCamel(s string) string {
...
+		y = v
+		if x >= 0 {
+			break
+		}
+	}
+	if x == -1 {
+		return s
+	}
+
+	if (x + len(y) - 1) == len(s)-1 {
```
Comment (estroz) [inline pkg/ansible/paramconv/paramconv.go:94]: I'd like to see some logic comments here too. It isn't clear what any of these conditions are doing.

File: pkg/ansible/paramconv/paramconv.go [hunk 18]
```diff
@@ -70,6 +70,49 @@ func ToCamel(s string) string {
 	return ret
 }
 
+// This function modifies the string to handle special words that are a part of wordMapping
+/* In this function we process the input string and find the occurrence of the values in wordMapping
+   and we append "_" on both sides of that word so that it is considered to be one word*/
+// We even handle plurals of the values in wordMapping
+// If we find "s" i.e plural, we capitalize it so that it is considered to be a part of same abbreviation
+
+func preprocessWordMapping(s string) string {
```
Comment (camilamacedo86) [inline pkg/ansible/paramconv/paramconv.go:79]: In go the comment should start // preprocessWordMapping ...

Following a suggestion.  (please check the example and adapt as you wish)

```suggestion
// preprocessWordMapping will check if the string informed contains special words mapped in the `wordMapping` // and its plurals and return it handled accordingly to do the snake. Note that, if the special word or its plural is 
// found the character `_` is appended to as prefixes and postfixes to the special world found. For example, if the
// string is `egressIP` the IP is a special word and then, the string   egress_ip will be returned. Also, beware that if
// the next character of the special world be an S( i.e plural ), it will be capitalized for the value be considered part
// of the same abbreviation.
func preprocessWordMapping(s string) string {
```

File: pkg/ansible/paramconv/paramconv.go [hunk 19]
```diff
@@ -70,6 +70,49 @@ func ToCamel(s string) string {
...
+	for _, v := range wordMapping {
+		x = strings.Index(s, v)
+		y = v
+		if x >= 0 {
+			break
+		}
+	}
+	if x == -1 {
+		return s
+	}
```
Comment (camilamacedo86) [inline pkg/ansible/paramconv/paramconv.go:92]: could not it be replaced by; 

```
       found := ""
	for _, word := range wordMapping {
		if strings.contains(value, word) {
                         found = word
                         break                          
                 } 
	}
	if found == ''" {
		return value
	}
} 
```

File: pkg/ansible/paramconv/paramconv.go [hunk 20]
```diff
@@ -70,6 +70,49 @@ func ToCamel(s string) string {
...
+/* In this function we process the input string and find the occurrence of the values in wordMapping
+   and we append "_" on both sides of that word so that it is considered to be one word*/
+// We even handle plurals of the values in wordMapping
+// If we find "s" i.e plural, we capitalize it so that it is considered to be a part of same abbreviation
+
+func preprocessWordMapping(s string) string {
+	var x int
+	var y string
+
+	for _, v := range wordMapping {
```
Comment (camilamacedo86) [inline pkg/ansible/paramconv/paramconv.go:83]: please use word instead of v here to make easier the understanding.

File: pkg/ansible/paramconv/paramconv.go (hunk 18 above)
Comment (camilamacedo86) [reply pkg/ansible/paramconv/paramconv.go:79]: ps IMO would be easier to understand if we rename s for value.

File: pkg/ansible/paramconv/paramconv.go [hunk 21]
```diff
@@ -70,6 +70,49 @@ func ToCamel(s string) string {
...
+			}
+		} else {
+			if x == 0 {
+				s = s[:(x+len(y))] + "_" + s[(x+len(y)):]
+			} else {
+				s = s[:x] + "_" + s[x:(x+len(y))] + "_" + s[(x+len(y)):]
+			}
+		}
+	}
+	return s
```
Comment (camilamacedo86) [inline pkg/ansible/paramconv/paramconv.go:113]: could not it be a code more readable by using the APi features from strings module or regex indeed ?

File: pkg/ansible/paramconv/paramconv.go [hunk 22]
```diff
@@ -84,6 +127,7 @@ func ToSnake(s string) string {
 	bits := []string{}
 	n := ""
 	iReal := -1
+	s = preprocessWordMapping(s)
```
Comment (camilamacedo86) [inline pkg/ansible/paramconv/paramconv.go:130]:
```suggestion
        // append underscore (_) as prefix and postfix to isolate special words defined in the wordMapping 
	s = preprocessWordMapping(s)
```

File: pkg/ansible/paramconv/paramconv.go [hunk 23]
```diff
@@ -70,6 +70,54 @@ func ToCamel(s string) string {
 	return ret
 }
 
+// preprocessWordMapping() will check if the string informed contains special words mapped in
```
Comment (estroz) [inline pkg/ansible/paramconv/paramconv.go:73]:
```suggestion
// preprocessWordMapping() will check if value contains special words mapped in
```

File: pkg/ansible/paramconv/paramconv.go [hunk 24]
```diff
@@ -70,6 +70,54 @@ func ToCamel(s string) string {
...
+// the 'wordMapping' and its plurals and is returned to ToSnake() for further processing.
```
Comment (estroz) [inline pkg/ansible/paramconv/paramconv.go:74]:
```suggestion
// wordMapping or is plural, then processes it such that ToSnake() can convert it to snake case.
```

File: pkg/ansible/paramconv/paramconv.go [hunk 25]
```diff
@@ -70,6 +70,54 @@ func ToCamel(s string) string {
...
+// Note that, if the special word or its plural is found the character "_" is appended to
```
Comment (estroz) [inline pkg/ansible/paramconv/paramconv.go:75]:
```suggestion
// If value contains a special word, the character "_" is appended
```

File: pkg/ansible/paramconv/paramconv.go [hunk 26]
```diff
@@ -70,6 +70,54 @@ func ToCamel(s string) string {
...
+// as prefixes and postfixes to the special world found. For example, if the string is "egressIP"
```
Comment (estroz) [inline pkg/ansible/paramconv/paramconv.go:76]:
```suggestion
// as a prefix and postfix to the special word found. For example, the string "egressIP", which
```

File: pkg/ansible/paramconv/paramconv.go [hunk 27]
```diff
@@ -70,6 +70,54 @@ func ToCamel(s string) string {
...
+// the IP is a special word and then, the string  egress_IP will be returned.
```
Comment (estroz) [inline pkg/ansible/paramconv/paramconv.go:77]:
```suggestion
// contains the special word "IP", will be returned as "egress_IP_".
```

File: pkg/ansible/paramconv/paramconv.go [hunk 28]
```diff
@@ -70,6 +70,54 @@ func ToCamel(s string) string {
...
+// Also, beware that if the next character of the special word be an "s" (i.e plural of the word
```
Comment (estroz) [inline pkg/ansible/paramconv/paramconv.go:78]:
```suggestion
// If the last character of the special word is an "s" (i.e plural of the word
```

File: pkg/ansible/paramconv/paramconv.go [hunk 29]
```diff
@@ -70,6 +70,54 @@ func ToCamel(s string) string {
...
+// found in 'wordMapping'), it will be capitalized to be considered part of the same abbreviation.
```
Comment (estroz) [inline pkg/ansible/paramconv/paramconv.go:79]:
```suggestion
// found in wordMapping), it is considered a part of that word and will be capitalized.
```

File: pkg/ansible/paramconv/paramconv.go [hunk 30]
```diff
@@ -70,6 +70,54 @@ func ToCamel(s string) string {
...
+		x = strings.Index(value, word)
+		y = word
+		if x >= 0 {
+			break
+		}
+	}
+	if x == -1 {
+		return value
+	}
+	// This is if the special non-plural word appears at the end of the string
```
Comment (estroz) [inline pkg/ansible/paramconv/paramconv.go:94]:
```suggestion
	// The special non-plural word appears at the end of the string.
```

File: pkg/ansible/paramconv/paramconv.go [hunk 31]
```diff
@@ -70,6 +70,54 @@ func ToCamel(s string) string {
...
+				value = value[:x] + "_" + value[x:(x+len(y))] + "S"
+			} else if x == 0 {
+				value = value[:(x+len(y))] + "S" + "_" + value[(x+len(y)+1):]
+			} else {
+				value = value[:x] + "_" + value[x:(x+len(y))] + "S" + "_" + value[(x+len(y)+1):]
+			}
+			// Under this else condition it handles the cases for non-plural words that come in Starting
+			//  and Middle of the string
+		} else {
+			if x == 0 {
```
Comment (estroz) [inline pkg/ansible/paramconv/paramconv.go:111]:
```suggestion
		} else {
			// The special non-plural word occurs at the start or somewhere in the middle of value.
			if x == 0 {
```

File: pkg/ansible/paramconv/paramconv.go [hunk 32]
```diff
@@ -70,6 +70,54 @@ func ToCamel(s string) string {
...
+		return value
+	}
+	// This is if the special non-plural word appears at the end of the string
+	if (x + len(y) - 1) == len(value)-1 {
+		value = value[:x] + "_" + value[x:]
+	} else {
+		// Under the following if: its the cases for handling plural words if the come in End, Starting
+		// and Middle respectively
+		if value[x+len(y)] == 's' {
+			if x+len(y) == len(value)-1 {
```
Comment (estroz) [inline pkg/ansible/paramconv/paramconv.go:101]:
```suggestion
		if value[x+len(y)] == 's' {
			// The special plural word occurs at the end, start, or somewhere in the middle of value.
			if x+len(y) == len(value)-1 {
```

File: changelog/fragments/3173-bugfix.yaml [hunk 33]
```diff
@@ -0,0 +1,6 @@
+entries:
+  - description: >
+      Handle variables with certain keywords.
+      Fix Ansible based-Operators for not mangled variables with uppercase characters when they have special words (e.g IP,HTTP, etc).
```
Comment (estroz) [inline changelog/fragments/3173-bugfix.yaml:4]:
```suggestion
      Prevent the ansible-operator from mangling variables containing the uppercase representations of special words (e.g IP, HTTP, etc).
```

File: pkg/ansible/paramconv/paramconv_test.go [hunk 34]
```diff
@@ -181,6 +186,41 @@ func TestToSnake(t *testing.T) {
 			args: args{"this_should_be_a_noop"},
 			want: "this_should_be_a_noop",
 		},
+		{
+			name: "should handle special words",
```
Comment (estroz) [inline pkg/ansible/paramconv/paramconv_test.go:190]:
```suggestion
			name: "should handle special plural word at end",
```

File: pkg/ansible/paramconv/paramconv_test.go [hunk 35]
```diff
@@ -181,6 +186,41 @@ func TestToSnake(t *testing.T) {
...
+			args: args{"egressIPs"},
+			want: "egress_ips",
+		},
+		{
+			name: "should handle special words",
```
Comment (estroz) [inline pkg/ansible/paramconv/paramconv_test.go:195]:
```suggestion
			name: "should handle special plural word in middle",
```

File: pkg/ansible/paramconv/paramconv_test.go [hunk 36]
```diff
@@ -181,6 +186,41 @@ func TestToSnake(t *testing.T) {
...
+			args: args{"egressIPsEgress"},
+			want: "egress_ips_egress",
+		},
+		{
+			name: "should handle special words",
```
Comment (estroz) [inline pkg/ansible/paramconv/paramconv_test.go:200]:
```suggestion
			name: "should handle special plural word in middle followed by lowercase letter",
```

File: pkg/ansible/paramconv/paramconv_test.go [hunk 37]
```diff
@@ -181,6 +186,41 @@ func TestToSnake(t *testing.T) {
...
+			args: args{"egressIPsegress"},
+			want: "egress_ips_egress",
+		},
+		{
+			name: "should handle special words",
```
Comment (estroz) [inline pkg/ansible/paramconv/paramconv_test.go:205]:
```suggestion
			name: "should handle special plural word at the start",
```

File: pkg/ansible/paramconv/paramconv_test.go [hunk 38]
```diff
@@ -181,6 +186,41 @@ func TestToSnake(t *testing.T) {
...
+			args: args{"IPsegress"},
+			want: "_ips_egress",
+		},
+		{
+			name: "should handle special words",
```
Comment (estroz) [inline pkg/ansible/paramconv/paramconv_test.go:210]:
```suggestion
			name: "should handle special word at the end",
```

File: pkg/ansible/paramconv/paramconv_test.go [hunk 39]
```diff
@@ -181,6 +186,41 @@ func TestToSnake(t *testing.T) {
...
+			args: args{"egressIP"},
+			want: "egress_ip",
+		},
+		{
+			name: "should handle special words",
```
Comment (estroz) [inline pkg/ansible/paramconv/paramconv_test.go:215]:
```suggestion
			name: "should handle special word in the middle",
```

File: pkg/ansible/paramconv/paramconv_test.go [hunk 40]
```diff
@@ -181,6 +186,41 @@ func TestToSnake(t *testing.T) {
...
+			args: args{"egressIPEgress"},
+			want: "egress_ip_egress",
+		},
+		{
+			name: "should handle special words",
```
Comment (estroz) [inline pkg/ansible/paramconv/paramconv_test.go:220]:
```suggestion
			name: "should handle special word in the middle followed by lowercase",
```

File: pkg/ansible/paramconv/paramconv.go [hunk 41]
```diff
@@ -70,6 +70,52 @@ func ToCamel(s string) string {
...
 }
 
+// preprocessWordMapping() will check if value contains special words mapped or its plural in
+// wordMapping, then processes it such that ToSnake() can convert it to snake case.
+// If value contains special word, the character "_" is appended as a prefix and postfix
+// to the special word found. For example, if the input string is "egressIP",
+// which contains is a special word "IP", the function will return "egress_IP".
+// If the last character of the special word is an "s" (i.e plural of the word
+// found in wordMapping), it is considered a part of that word and will be capitalized.
+func preprocessWordMapping(value string) string {
```
Comment (estroz) [inline pkg/ansible/paramconv/paramconv.go:80]: Can a value contain multiple special words?

Comment (Venkat19967) [reply pkg/ansible/paramconv/paramconv.go:80]: No it does not address multiple special words. Would you want me to add that feature?

Comment (estroz) [reply pkg/ansible/paramconv/paramconv.go:80]: @fabianvf can an ansible variable name contain multiple special words?

File: changelog/fragments/3173-bugfix.yaml [hunk 42]
```diff
@@ -0,0 +1,5 @@
+entries:
+  - description: >
+      Prevent the ansible-operator from mangling variables containing the uppercase representations of special words (e.g IP, HTTP, etc).
+    kind: "bugfix"
+    breaking: false
```
Comment (fabianvf) [inline changelog/fragments/3173-bugfix.yaml:5]: I would actually flag this as a potentially breaking change, as the parameters being passed into the playbook are potentially changing

File: pkg/ansible/paramconv/paramconv.go (hunk 41 above)
Comment (fabianvf) [reply pkg/ansible/paramconv/paramconv.go:80]: Yeah, can be any arbitrary string

File: pkg/ansible/paramconv/paramconv.go [hunk 43]
```diff
@@ -70,6 +70,49 @@ func ToCamel(s string) string {
 	return ret
 }
 
+// preprocessWordMapping() will check if value contains special words mapped or its plural in
+// wordMapping, then processes it such that ToSnake() can convert it to snake case.
+// If value contains special word, the character "_" is appended as a prefix and postfix
+// to the special word found. For example, if the input string is "egressIP",
+// which contains is a special word "IP", the function will return "egress_IP".
```
Comment (jmrodri) [inline pkg/ansible/paramconv/paramconv.go:77]: nit: remove extra 'is'. line should read: "which contains a special word..."

File: changelog/fragments/3173-bugfix.yaml [hunk 44]
```diff
@@ -0,0 +1,10 @@
+entries:
+  - description: >
+      Prevent the ansible-operator from mangling variables containing the uppercase representations of special words (e.g IP, HTTP, etc).
```
Comment (jmrodri) [inline changelog/fragments/3173-bugfix.yaml:3]: nit: ansible-operator should be "Ansible Operator"

File: changelog/fragments/3173-bugfix.yaml [hunk 45]
```diff
@@ -0,0 +1,10 @@
...
+    kind: change
+    breaking: true
+
+    migration:
+      header: This new change will fix the ansible variables that are getting mangled
```
Comment (jmrodri) [inline changelog/fragments/3173-bugfix.yaml:8]: nit: ansible should be capitalized: Ansible

File: changelog/fragments/3173-bugfix.yaml [hunk 46]
```diff
@@ -0,0 +1,10 @@
...
+      body: >
+        _/operator-sdk/pkg/ansible/paramconv/paramconv.go will be able to handle variables properly(Camel and Snake case conversion)
```
Comment (jmrodri) [inline changelog/fragments/3173-bugfix.yaml:10]: nit: add a space after properly -> "properly (Camel and Snake)

File: pkg/ansible/paramconv/paramconv.go [hunk 47]
```diff
@@ -70,6 +70,49 @@ func ToCamel(s string) string {
...
+				} else {
+					if x == 0 {
+						value = value[:(x+len(y))] + "_" + value[(x+len(y)):]
+					} else {
+						value = value[:x] + "_" + value[x:(x+len(y))] + "_" + value[(x+len(y)):]
+					}
+				}
+			}
+		}
+	}
```
Comment (jmrodri) [inline pkg/ansible/paramconv/paramconv.go:111]: I was able to remove all the extra `else` statements. The tests pass. I'm not 100% sure if the logic is the same yet. 

```golang
    for _, word := range wordMapping {                                             
        idx = strings.Index(value, word)                                           
        if idx >= 0 {                                                              
            // The special non-plural word appears at the end of the string.       
            if (idx + len(word) - 1) == len(value)-1 {                             
                value = value[:idx] + "_" + value[idx:]                            
            } else if value[idx+len(word)] == 's' {                                
                // The special plural word occurs at the end, start, or somewhere in the middle of value.
                if idx+len(word) == len(value)-1 {                                 
                    value = value[:idx] + "_" + value[idx:(idx+len(word))] + "S" 
                } else if idx == 0 {                                               
                    value = value[:(idx+len(word))] + "S" + "_" + value[(idx+len(word)+1):]
                } else {                                                           
                    value = value[:idx] + "_" + value[idx:(idx+len(word))] + "S" + "_" + value[(idx+len(word)+1):]
                }                                                                  
            } else if idx == 0 {                                                   
                // The special non-plural word occurs at the start or somewhere in the middle of value.
                value = value[:(idx+len(word))] + "_" + value[(idx+len(word)):] 
            } else {                                                               
                value = value[:idx] + "_" + value[idx:(idx+len(word))] + "_" + value[(idx+len(word)):]
            }                                                                      
        }                                                                          
    } 
```

Comment (Venkat19967) [reply pkg/ansible/paramconv/paramconv.go:111]: I just checked the logic. It should work fine. let me update the code with this and run some tests.

Comment (Venkat19967) [reply pkg/ansible/paramconv/paramconv.go]: I have implemented this logic (Removed the unnecessary else{} and eliminated the use of one extra variable

File: changelog/fragments/3173-bugfix.yaml [hunk 48]
```diff
@@ -0,0 +1,10 @@
+entries:
+  - description: >
+      Prevent the ansible-operator from mangling variables containing the uppercase representations of special words (e.g IP, HTTP, etc).
+    kind: change
+    breaking: true
+
+    migration:
+      header: This new change will fix the ansible variables that are getting mangled
+      body: >
+        _/operator-sdk/pkg/ansible/paramconv/paramconv.go will be able to handle variables properly (Camel and Snake case conversion)
```
Comment (fabianvf) [inline changelog/fragments/3173-bugfix.yaml:10]: This should contain the instructions for a user to avoid being broken by this change. It shouldn't mention any internal code (users don't know or care about this file), and it should have a few examples of how variables will be handled differently

File: changelog/fragments/3173-bugfix.yaml [hunk 49]
```diff
@@ -0,0 +1,17 @@
+entries:
+  - description: >
+      Prevent the ansible-operator from mangling variables containing the uppercase representations of special words (e.g IP, HTTP, etc).
+    kind: change
+    breaking: true
```
Comment (estroz) [inline changelog/fragments/3173-bugfix.yaml:5]: Is this really breaking? Does an operator author have to change anything in an existing project?

Comment (fabianvf) [reply changelog/fragments/3173-bugfix.yaml:5]: Yeah, it could be changing the parameters that are passed in, which means they could be referencing a variable that no longer exists. The only workaround for this bug was to reference the improperly converted variable so I expect anyone that hit it and continued going will be affected.

**Comment by tengqm** [comment]
the implementation is kinda difficult to maintain ... because it is considering many unrealistic cases.
maybe a simplified implementation would be something like this:

```
import (
	"regexp"
	"strings"
)

func camelToSnake(s string) string {
	re := regexp.MustCompile(`([a-z0-9])([A-Z])`)
	res := re.ReplaceAllString(s, `${1}_${2}`)
	return strings.ToLower(res)
}
```
